File: app/api/sql_templates.py
"""SQL 工作台 v0.4 模板库 endpoints。

URL 风格跟同期 SQL 工作台保持一致(`/api/sql-templates/...`)。

editor+ 才能 POST/PUT/DELETE,viewer 只读列表 + 详情(GET)。内置模板 (id 前缀
`builtin:`) 永远在响应里,但 update / delete 时 store 自己拒,返回 403。

import/export 是显式两个 endpoint,前端"上传 JSON" / "下载 JSON" 按钮直接调。
"""
from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, Body, Depends, HTTPException, Query

from app.models import SQLTemplate, SQLTemplateCreate, SQLTemplateUpdate, User
from app.services.auth import require_role
from app.sqlide.template_store import sql_template_store

logger = logging.getLogger(__name__)


# 全局 viewer+:列表 / 详情可读;具体 POST/PUT/DELETE 用 require_role("editor")
router = APIRouter(dependencies=[Depends(require_role("viewer"))])


def _parse_csv(v: str) -> list[str]:
    return [x.strip() for x in v.split(",") if x.strip()] if v else []


@router.get("/api/sql-templates")
def list_templates(
    q: str = Query(default=""),
    tag: str = Query(default="", description="逗号分隔多 tag,AND 命中"),
    db_type: str = Query(default=""),
    project_id: str | None = Query(default=None),
) -> dict[str, Any]:
    """列模板,带过滤。返回 {items, count}。"""
    items = sql_template_store.list(
        q=q, tags=_parse_csv(tag) or None,
        db_type=db_type, project_id=project_id,
    )
    return {"items": [t.model_dump(mode="json") for t in items], "count": len(items)}


@router.get("/api/sql-templates/{template_id}")
def get_template(template_id: str) -> dict[str, Any]:
    t = sql_template_store.get(template_id)
    if not t:
        raise HTTPException(404, "template not found")
    return t.model_dump(mode="json")


@router.post("/api/sql-templates")
def create_template(
    payload: SQLTemplateCreate = Body(...),
    current: User = Depends(require_role("editor")),
) -> dict[str, Any]:
    t = sql_template_store.create(payload, created_by=current.id)
    return t.model_dump(mode="json")


@router.put("/api/sql-templates/{template_id}")
def update_template(
    template_id: str,
    payload: SQLTemplateUpdate = Body(...),
    current: User = Depends(require_role("editor")),
) -> dict[str, Any]:
    try:
        t = sql_template_store.update(template_id, payload)
    except PermissionError as e:
        raise HTTPException(403, str(e))
    except KeyError:
        raise HTTPException(404, "template not found")
    return t.model_dump(mode="json")


@router.delete("/api/sql-templates/{template_id}")
def delete_template(
    template_id: str,
    current: User = Depends(require_role("editor")),
) -> dict[str, Any]:
    try:
        sql_template_store.delete(template_id)
    except PermissionError as e:
        raise HTTPException(403, str(e))
    except KeyError:
        raise HTTPException(404, "template not found")
    return {"ok": True, "id": template_id}


@router.post("/api/sql-templates/import")
def import_templates(
    payload: dict = Body(..., description="{templates: [...], overwrite_by_name?: bool}"),
    current: User = Depends(require_role("editor")),
) -> dict[str, Any]:
    items = payload.get("templates") or []
    if not isinstance(items, list):
        raise HTTPException(400, "templates 必须是 list")
    overwrite = bool(payload.get("overwrite_by_name", False))
    report = sql_template_store.import_templates(
        items, created_by=current.id, overwrite_by_name=overwrite,
    )
    return {"ok": True, **report}


@router.get("/api/sql-templates/export/json")
def export_templates(
    include_builtin: bool = Query(default=False),
    project_id: str | None = Query(default=None),
) -> dict[str, Any]:
    """返 {templates: [...]} 让前端直接下载为 .json 文件。"""
    items = sql_template_store.export(include_builtin=include_builtin, project_id=project_id)
    return {"templates": items, "count": len(items)}

File: frontend/frontend/src/stores/sqlTemplates.ts
/**
 * SQL 工作台 v0.4 — 模板库 store。
 *
 * 单独建一个 store(不塞 sqlWorkbench)是因为关注分离:
 *   - sqlWorkbench:console / execute / metadata,运行时数据
 *   - sqlTemplates:模板 CRUD,准静态配置数据
 *
 * 数据流:
 *   1. SqlWorkbenchView onMounted → loadTemplates()
 *   2. 用户点"保存为模板" → createTemplate(payload)
 *   3. 用户在模板 tab 编辑 → updateTemplate(id, payload)
 *   4. 用户删除 → deleteTemplate(id)
 *   5. 用户导入 JSON → importTemplates(list, overwrite)
 *   6. 用户点导出 → exportTemplates() 拿 JSON 返回给浏览器下载
 */
import { reactive, ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { apiGet, apiJson } from '../api'

export type RiskLevel = 'low' | 'medium' | 'high'

export interface SQLTemplate {
  id: string
  name: string
  description: string
  tags: string[]
  db_types: string[]
  project_id: string
  risk_level: RiskLevel
  sql: string
  created_by: string
  created_at: string
  updated_at: string
  builtin: boolean
}

export interface SQLTemplateCreate {
  name: string
  description?: string
  tags?: string[]
  db_types?: string[]
  project_id?: string
  risk_level?: RiskLevel
  sql: string
}

export interface ImportReport {
  ok: boolean
  created: number
  skipped: number
  errors: number
}

export const useSqlTemplatesStore = defineStore('sqlTemplates', () => {
  const templates = ref<SQLTemplate[]>([])
  const loading = ref(false)
  const lastError = ref('')

  // 过滤态 —— 用户在模板 panel 调,会触发 loadTemplates 重拉(简单方案;
  // 也可以前端纯客户端过滤,但 GET 时塞 query 更省内存且支持 server-side 排序)。
  const filters = reactive<{
    q: string
    tag: string         // 逗号分隔
    db_type: string
    project_id: string | null  // null 表示"不过滤";空串表示"只看全局"
  }>({ q: '', tag: '', db_type: '', project_id: null })

  // 派生:builtin / user 两类计数,UI 展示用
  const builtinCount = computed(() => templates.value.filter(t => t.builtin).length)
  const userCount = computed(() => templates.value.filter(t => !t.builtin).length)

  async function loadTemplates(): Promise<void> {
    loading.value = true
    lastError.value = ''
    try {
      const qs = new URLSearchParams()
      if (filters.q) qs.set('q', filters.q)
      if (filters.tag) qs.set('tag', filters.tag)
      if (filters.db_type) qs.set('db_type', filters.db_type)
      if (filters.project_id !== null) qs.set('project_id', filters.project_id)
      const resp = await apiGet<{ items: SQLTemplate[]; count: number }>(
        `/api/sql-templates${qs.toString() ? '?' + qs.toString() : ''}`,
      )
      templates.value = resp.items || []
    } catch (e: unknown) {
      lastError.value = (e as Error)?.message || String(e)
    } finally {
      loading.value = false
    }
  }

  async function createTemplate(payload: SQLTemplateCreate): Promise<SQLTemplate> {
    const t = await apiJson<SQLTemplate>('/api/sql-templates', 'POST', payload)
    templates.value.unshift(t)
    return t
  }

  async function updateTemplate(id: string, payload: SQLTemplateCreate): Promise<SQLTemplate> {
    const t = await apiJson<SQLTemplate>(`/api/sql-templates/${id}`, 'PUT', payload)
    const idx = templates.value.findIndex(x => x.id === id)
    if (idx >= 0) templates.value[idx] = t
    return t
  }

  async function deleteTemplate(id: string): Promise<void> {
    await apiJson(`/api/sql-templates/${id}`, 'DELETE')
    templates.value = templates.value.filter(t => t.id !== id)
  }

  async function importTemplates(items: any[], overwriteByName: boolean = false): Promise<ImportReport> {
    const report = await apiJson<ImportReport>('/api/sql-templates/import', 'POST', {
      templates: items, overwrite_by_name: overwriteByName,
    })
    // 拉一次最新列表
    await loadTemplates()
    return report
  }

  async function exportTemplates(includeBuiltin: boolean = false): Promise<any[]> {
    const qs = new URLSearchParams()
    qs.set('include_builtin', includeBuiltin ? 'true' : 'false')
    const resp = await apiGet<{ templates: any[]; count: number }>(
      `/api/sql-templates/export/json?${qs.toString()}`,
    )
    return resp.templates || []
  }

  return {
    templates, loading, lastError, filters,
    builtinCount, userCount,
    loadTemplates, createTemplate, updateTemplate, deleteTemplate,
    importTemplates, exportTemplates,
  }
})
